builtins: use Youngs-Crammer algorithm for aggregation functions

mneverov opened the pull request:
builtins: use Youngs-Crammer algorithm for aggregation functions

This commit replaces existing algorithm for correlation calculation with the Youngs-Crammer algorithm ported from Postgresql to reduce rounding errors. It introduces the base structure for aggregate functions for statistics.
It also amends aggregates builtin tests so functions with several arguments can be tested.

Release note: None

Relates to #41274

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=55268) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=55268) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=55268) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55268)
<!-- Reviewable:end -->

rohany commented:
cc @yuzefovich

mneverov commented:
The tests for multiple nodes and distributed sql mode "on" are failing because correlation calculation depends on the order of the values.
This is true for the old method as well as for the Youngs-Crammer algorithm ported from the Postgres.

Consider the following example:

```
-- Postgres

CREATE TABLE aggtest
(
    a double precision,
    b double precision
);

insert into aggtest (a, b)
values (0, 0.09561),
       (12, 124.78),
       (42, 324.78),
       (42, 324.78),
       (42, 384.78),
       (42, 324.78),
       (56, 7.8),
       (56, 7.8),
       (56, 7.8),
       (56, 7.8),
       (56, 324.78),
       (87, 127.001),
       (100, 99.097),
       (100, 99.097),
       (100, 99.097);

SELECT corr(b, a)
FROM aggtest;

-- PG:             -0.14162191454732073
-- Old algorithm:  -0.14162191454732112
-- Youngs-Crammer: -0.14162191454732073

truncate aggtest;

-- the same values in different order
insert into aggtest (a, b)
values (56, 7.8),
       (56, 7.8),
       (42, 324.78),
       (56, 7.8),
       (100, 99.097),
       (42, 324.78),
       (56, 7.8),
       (100, 99.097),
       (100, 99.097),
       (42, 324.78),
       (56, 324.78),
       (42, 384.78),
       (12, 124.78),
       (0, 0.09561),
       (87, 127.001)
;

SELECT corr(b, a)
FROM aggtest;

-- PG:             -0.1416219145473208
-- Old algorithm:  -0.14162191454732093
-- Youngs-Crammer: -0.1416219145473208
```

My understanding, based on the [design doc](https://github.com/cockroachdb/cockroach/blob/master/docs/design.md#logical-planning), is that it's possible to define an order on the output stream, so that intervals from different nodes would merge as if the operation was executed locally (i.e. rows will be ordered based on ordinal number). 

@yuzefovich could you please confirm that my understanding is correct and advice how I can define the order on the `corr` op?

yuzefovich reviewed (commented):
Thanks for taking this on and for the thorough investigation of this precision issue! I agree with your assessment that the order of float operands seems to matter, so we need to think of some remedy.

> My understanding, based on the design doc, is that it's possible to define an order on the output stream, so that intervals from different nodes would merge as if the operation was executed locally (i.e. rows will be ordered based on ordinal number).

I think your understanding is correct, but I don't like the idea of requiring some order in order to get the deterministic result when talking about such tiny precision (beyond 15 digits as documented by [Postgres](https://www.postgresql.org/docs/9.0/datatype-numeric.html)). I'll confirm with my colleagues tomorrow, but I think we should just modify the tests to allow for some deviation, i.e. instead of something like
```
query R
SELECT corr(y, x)::decimal FROM statistics_agg_test
----
0.045228963191363145
```
we would have something like
```
query B
SELECT abs(corr(y, x)::decimal-0.045228963191363145) < power(10, -15) FROM statistics_agg_test
----
true
```
My hesitation of imposing an ordering comes from the fact that it'll likely have performance implications (we might need to plan a sort before an aggregator), and we're already satisfying the precision requirements mentioned by Postgres. I believe some non-determinism is acceptable in this case. What do you think?

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-:-MJUXuhi2WkskQPQVnHR:b47rbc1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUWW0M5J2Dgzjb1Mxo:-MJUWW0M5J2Dgzjb1Mxp:b-cumuxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1739)):*
> ```Go
> }
> 
> /*
> ```

nit: we tend to use `//` for all comments, not `/*` and `*/`.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1758 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUWgDr4CnCurjQW1pN:-MJUWgDr4CnCurjQW1pO:bi0smtw) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1758)):*
> ```Go
>  * to minimize code space instead.
>  */
> ```

nit: we usually don't have an empty line between a comment and a struct.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUX_uPAmiyWubn1Mz7:-MJUX_uPAmiyWubn1Mz8:b-f006wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1759)):*
> ```Go
>  */
> 
> type regressionAccumulate struct {
> ```

super nit: I'd probably do `s/accumulate/accumulator/`.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUXByS9qXDWiP3xui9:-MJUXByS9qXDWiP3xuiA:blm8tzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1892)):*
> ```Go
> 	}
> 
> 	err = a.regAcc.add(y, x)
> ```

nit: could combine these two lines into one.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1910 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUXJjV1b9hrZPVro8p:-MJUXJjV1b9hrZPVro8q:b-o0ld8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1910)):*
> ```Go
> // Reset implements tree.AggregateFunc interface.
> func (a *corrAggregate) Reset(context.Context) {
> 	a.regAcc.n = 0
> ```

nit: could replace all the lines with this
```
a.regAcc = regressionAccumulate{}
```


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Another idea that was suggested by @asubiotto is to adjust the logic test harness itself in order to introduce a separate type marker for floats (separate from decimals) that allows for 10e-15 deviation, I'll look into that shortly.

yuzefovich commented:
I opened up #55530, and I would be interested if you could try it out (note that you'll need to modify the `query` invocation to use `F` type instead of `R` for floating point numbers that are non-deterministic).

mneverov commented:
> I opened up #55530, and I would be interested if you could try it out (note that you'll need to modify the `query` invocation to use `F` type instead of `R` for floating point numbers that are non-deterministic).

@yuzefovich thanks!
Just tried it and all logic tests for aggregate testcase pass. Will update this PR when #55530 is merged into the master.

> My hesitation of imposing an ordering comes from the fact that it'll likely have performance implications (we might need to plan a sort before an aggregator), and we're already satisfying the precision requirements mentioned by Postgres. I believe some non-determinism is acceptable in this case. What do you think?

I haven't ever needed 16 digits precision, so I'm ok with this. Maybe, it is worth to mention the deviation in the docs?

yuzefovich commented:
#55530 has just merged.

> I haven't ever needed 16 digits precision, so I'm ok with this. Maybe, it is worth to mention the deviation in the docs?

I think that floating point precision is a pretty confusing topic, and talking about it in the docs might only add to the confusion. I'd expect most people to not care whether there is non-determinism (depending on the distribution of the execution) in the result with `float` type, so I'd probably not call it out.

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55268)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUX_uPAmiyWubn1Mz7:-MJp-JAq9p_BMu_ihILK:b-nfge6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1759)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: I'd probably do `s/accumulate/accumulator/`.
</blockquote></details>

sorry didn't get it, could you please explain this one if it is still relevant with the latest changes?
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUXByS9qXDWiP3xui9:-MJp-_464_zVnNhWg6-j:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1892)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: could combine these two lines into one.
</blockquote></details>

:+1: 


<!-- Sent from Reviewable.io -->

mneverov commented:
@yuzefovich could you please have another look?

The PR got quite large so I split it in two: this one ports the algorithm from Postgres to reuse in other aggregation functions later. I found some tests for `corr` missing.
The actual implementation of `covar_pop` I can do as a separate PR or an another commit to this one and then squash after the review.

yuzefovich reviewed (approved):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some nits.

I think it's reasonable to have this PR as a standalone, and then work on `covar_pop` in a different PR.

nit: you should update the PR title to the commit title.

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-:-MJraEzr9HRbfaWgBQf8:btcv1b2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mneverov)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUX_uPAmiyWubn1Mz7:-MJrYyICCAXs_r28LcMW:b7jqern) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1759)):*
<details><summary><i>Previously, mneverov (Max Neverov) wrote…</i></summary><blockquote>

sorry didn't get it, could you please explain this one if it is still relevant with the latest changes?
</blockquote></details>

The notation `s/abc/xyz/` means "replace `abc` with `xyz`" (comes from `sed` command).

My nit comes from the fact that I think that "accumulate" is not a noun (unlike "aggregate"), and for the struct names we prefer nouns, so `regressionAccumulator` or `regressionAccumulatorBase` would sound better to me.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJrZpiO8fkhr162YcaZ:-MJrZpiO8fkhr162Yca_:b-q5sr5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b24abc4b67c0514e581b24111cb3aae857a0ae3/pkg/sql/sem/builtins/aggregate_builtins.go#L1789)):*
> ```Go
> // Reset implements tree.AggregateFunc interface.
> func (a *regressionAccumulateBase) Reset(context.Context) {
> 	a.n = 0
> ```

nit: all these lines could be replaced with `*a = regressionAccumulateBase{}`.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1802 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJrZx8-EyhNMa-8ve1R:-MJrZx8-EyhNMa-8ve1S:bbhs689) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b24abc4b67c0514e581b24111cb3aae857a0ae3/pkg/sql/sem/builtins/aggregate_builtins.go#L1802)):*
> ```Go
> // Size implements tree.AggregateFunc interface.
> func (a *regressionAccumulateBase) Size() int64 {
> 	return sizeOfCorrAggregate
> ```

nit: we should define a separate `sizeOfRegressionAccumulateBase` and use it here.
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJr_vNB1ghCK8-TvEOp:-MJr_vNB1ghCK8-TvEOq:b-f9rjc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b24abc4b67c0514e581b24111cb3aae857a0ae3/pkg/sql/sem/builtins/aggregate_builtins_test.go#L75)):*
> ```Go
> }
> 
> func pivotArgs(args ...[]tree.Datum) [][]tree.Datum {
> ```

nit: why is this function called "pivot"? It seems to me it's more like "flattenArgs".


<!-- Sent from Reviewable.io -->

mneverov committed:
builtins: use Youngs-Crammer algorithm for aggregation functions

This commit replaces existing algorithm for correlation calculation with the Youngs-Crammer algorithm ported from Postgresql to reduce rounding errors. It introduces the base structure for aggregate functions for statistics.
It also amends aggregates builtin tests so functions with several arguments can be tested.

Release note: None

mneverov reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55268)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJUX_uPAmiyWubn1Mz7:-MJreJ2lB7GGTDNTh1Fy:b-guxjgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/501f8df878b2b62ea4f0c43cc61465c73a107c53/pkg/sql/sem/builtins/aggregate_builtins.go#L1759)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

The notation `s/abc/xyz/` means "replace `abc` with `xyz`" (comes from `sed` command).

My nit comes from the fact that I think that "accumulate" is not a noun (unlike "aggregate"), and for the struct names we prefer nouns, so `regressionAccumulator` or `regressionAccumulatorBase` would sound better to me.
</blockquote></details>

got it, thanks. 
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJrZpiO8fkhr162YcaZ:-MJreMgg1ErdUXn2nyHw:bla2j1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b24abc4b67c0514e581b24111cb3aae857a0ae3/pkg/sql/sem/builtins/aggregate_builtins.go#L1789)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: all these lines could be replaced with `*a = regressionAccumulateBase{}`.
</blockquote></details>

missed that, fixed, thanks
___
*[pkg/sql/sem/builtins/aggregate_builtins_test.go, line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-MJr_vNB1ghCK8-TvEOp:-MJrevIS3hD66I-nEqfv:b-x9dy9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b24abc4b67c0514e581b24111cb3aae857a0ae3/pkg/sql/sem/builtins/aggregate_builtins_test.go#L75)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: why is this function called "pivot"? It seems to me it's more like "flattenArgs".
</blockquote></details>

Thanks, "flattenArgs" is indeed describes better what it does


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
bors r=yuzefovich

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55268#-:-MJs33Q47aO6SarxMkNz:b-ewwdc)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->
